### ISSUE TITLE
slow rank-Z test/NA handling in factor variables

### ISSUE THREAD

**Issue by bbolker** [lme4/lme4#33]
From Kevin Thorpe.

Synthesized example:

```
## number of levels with each level of replication
levs <- c(800,300,150,100,50,50,50,20,20,5,2,2,2,2)
n <- seq_along(levs)
flevels <- seq(sum(levs))
set.seed(101)
fakedat <- data.frame(DA=factor(rep(flevels,rep(n,levs))),
                      zbmi=rnorm(sum(n*levs)))
fakedat[sample(nrow(fakedat),100),"zbmi"] <- NA
fakedat[sample(nrow(fakedat),100),"DA"] <- NA
library("lme4")
system.time(m5 <- lmer(zbmi ~ (1|DA) , data = fakedat))
system.time(m5 <- lmer(zbmi ~ (1|DA) , data = na.omit(fakedat)))
```

  This test reveals three problems:
- the source tarball on `lme4.r-forge.r-project.org/repos` , which is out of date, wrongly returns zero variance for `DA`
- the most current development lme4 version doesn't handle NAs in the factor variable properly (it throws an error, which is not the _worst_ thing in the world, but it should properly omit them)
- the same version is **very** slow because it tries to test the rank of Zt: the result is that the fit is very slow -- between 7 and 18 seconds depending on caching, etc etc, instead of 0.1 seconds (!!)

In particular, the test is 

```
    if ((rankZ <- rankMatrix(reTrms$Zt)) > nrow(fr)) {
        stop("rank(Z)>number of observations; variance-covariance matrix will be unidentifiable")
    }
```

this is quite slow, no matter which rankMatrix() method I tried (only about 7 seconds on the raw matrix -- 

   Suggestions?  Should we

(1) turn off the test completely? (for what it's worth I have found other pathological/overidentified cases that this test does _not_ catch)
(2) turn off the test above some matrix dimensions?  (if so, what dimensions?)
(3) in conjunction with #1 or #2, provide a control argument for turning the test off and on?

 Bottom line is that all this stuff is fixable, but it would be good to think a bit about how best to do it.
